src/main/java/org/spongepowered/api/item/recipe/crafting/CraftingRecipe.java

```diff
@@ -0,0 +1,100 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.recipe.crafting;
+
+import org.spongepowered.api.item.ItemTypes;
+import org.spongepowered.api.item.inventory.ItemStack;
+import org.spongepowered.api.item.inventory.type.GridInventory;
+import org.spongepowered.api.item.recipe.Recipe;
+import org.spongepowered.api.world.World;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * A CraftingRecipe represents some craftable recipe in the game.
+ *
+ * <p>It is essentially a Predicate that checks for if a recipe is valid as well
+ * as a function from a crafting matrix to a list of {@link ItemStack} (the
+ * crafting result), therefore making it an immutable interface.</p>
+ *
+ * <p>The passed in ItemGrid is usually a crafting inventory, e.g. a 2x2 or 3x3
+ * crafting matrix.</p>
+ *
+ * <p>The requirements of a CraftingRecipe can be general, they just have to
+ * eventually return a boolean given an itemgrid.</p>
+ */
+public interface CraftingRecipe extends Recipe {
+
+    /**
+     * Checks if the given {@link GridInventory} fits the required constraints
+     * to craft this {@link CraftingRecipe}.
+     *
+     * @param grid The ItemGrid to check for validity
+     * @param world The world this recipe would be used in
+     * @return True if the given input matches this recipe's requirements
+     */
+    boolean isValid(GridInventory grid, World world);
+
+    /**
+     * This method is preferred over the {@link #getExemplaryResult()}
+     * method, as it customizes the result further depending on the specified
+     * input {@param grid}.
+     *
+     * <p>It is advised to use the output of {@link #getExemplaryResult()},
+     * modify it accordingly, and {@code return} it.</p>
+     *
+     * @param grid The crafting input, typically 3x3 or 2x2
+     * @param world The world this recipe would be used in
+     * @return An {@link ItemStack} or {@link Optional#empty()} if the given
+     *         {@link GridInventory} does not match this recipe's requirements.
+     */
+    Optional<ItemStack> getResult(GridInventory grid, World world);
+
+    /**
+     * A list of items to be added to the inventory of the player when they
+     * craft the result. For example, if a player crafts a
+     * {@link ItemTypes#CAKE}, the empty buckets are returned to their
+     * inventory.
+     *
+     * @param grid The crafting input, typically 3x3 or 2x2
+     * @param world The world this recipe would be used in
+     * @return The list of items to be added to the inventory of the player
+     *         when the recipe has been fulfilled (possibly empty),
+     *         or {@link Optional#empty()} if the given
+     *         {@link GridInventory} does not match this recipe's requirements.
+     */
+    Optional<List<ItemStack>> getRemainingItems(GridInventory grid, World world);
```

Comment:
I quite unsure about this one. 

Wouldn't be a `Optional<CraftingResult>` better here?
Because currently you have to check the validity of the recipe twice (one for the result stack and one time for the remaining stacks).

If you use a `CraftingResult` which would store both of them at once you could skip one validity execution.